File: android/app/src/main/java/com/homebudgeting/data/BudgetRepository.kt
package com.homebudgeting.data

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

class BudgetRepository(private val storage: BudgetStorage) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val mutex = Mutex()
    private val _state = MutableStateFlow(BudgetState.Empty)
    val state: StateFlow<BudgetState> = _state.asStateFlow()

    init {
        scope.launch {
            _state.value = storage.load()
        }
    }

    suspend fun refresh() {
        mutex.withLock {
            _state.value = storage.load()
        }
    }

    suspend fun createMonth(monthKey: String) {
        updateState { current ->
            if (current.months.containsKey(monthKey)) return@updateState current
            val lastKey = current.months.keys.sorted().lastOrNull()
            val template = BudgetMonth()
            val newMonth = if (lastKey != null) {
                val lastMonth = current.months[lastKey]
                template.copy(
                    categories = lastMonth?.categories?.mapValues { (_, cat) -> cat.copy() } ?: emptyMap()
                )
            } else {
                template
            }
            current.copy(months = (current.months + (monthKey to newMonth)).toSortedMap())
        }
    }

    suspend fun deleteMonth(monthKey: String) {
        updateState { current ->
            if (!current.months.containsKey(monthKey)) return@updateState current
            val newMonths = current.months.toMutableMap()
            newMonths.remove(monthKey)
            current.copy(months = newMonths.toSortedMap(), ui = clearCollapsed(current.ui, monthKey))
        }
    }

    suspend fun upsertMonth(monthKey: String, transform: (BudgetMonth) -> BudgetMonth) {
        updateState { current ->
            val month = current.months[monthKey] ?: BudgetMonth()
            val updated = transform(month).ensureIntegrity()
            val newMonths = current.months.toMutableMap()
            newMonths[monthKey] = updated
            current.copy(months = newMonths.toSortedMap())
        }
    }

    suspend fun setMapping(mapping: PredictionMapping) {
        updateState { current -> current.copy(mapping = mapping.ensureIntegrity()) }
    }

    suspend fun setDescriptionMap(descMap: DescriptionMap) {
        updateState { current -> current.copy(descMap = descMap.ensureIntegrity()) }
    }

    suspend fun setDescriptionList(list: List<String>) {
        updateState { current -> current.copy(descList = list.distinctBy { it.lowercase() }) }
    }

    suspend fun setNotes(notes: List<Note>) {
        updateState { current -> current.copy(notes = notes.sortedByDescending { it.time }) }
    }

    suspend fun setCollapsed(monthKey: String, group: String, collapsed: Boolean) {
        updateState { current ->
            val collapsedMap = current.ui.collapsed.toMutableMap()
            val groupMap = (collapsedMap[monthKey] ?: emptyMap()).toMutableMap()
            if (collapsed) {
                groupMap[group] = true
            } else {
                groupMap.remove(group)
            }
            if (groupMap.isEmpty()) {
                collapsedMap.remove(monthKey)
            } else {
                collapsedMap[monthKey] = groupMap
            }
            current.copy(ui = UiPreferences(collapsed = collapsedMap))
        }
    }

    suspend fun setAllCollapsed(monthKey: String, groups: Collection<String>, collapsed: Boolean) {
        updateState { current ->
            val collapsedMap = current.ui.collapsed.toMutableMap()
            if (collapsed) {
                collapsedMap[monthKey] = groups.associateWith { true }
            } else {
                collapsedMap.remove(monthKey)
            }
            current.copy(ui = UiPreferences(collapsed = collapsedMap))
        }
    }

    suspend fun importData(incoming: BudgetState) {
        updateState { current -> mergeStates(current, incoming.ensureIntegrity()) }
    }

    private fun clearCollapsed(ui: UiPreferences, monthKey: String): UiPreferences {
        if (!ui.collapsed.containsKey(monthKey)) return ui
        val map = ui.collapsed.toMutableMap()
        map.remove(monthKey)
        return UiPreferences(collapsed = map)
    }

    private suspend fun updateState(block: (BudgetState) -> BudgetState) {
        mutex.withLock {
            val updated = block(_state.value).ensureIntegrity()
            storage.save(updated)
            _state.value = updated
        }
    }

    private fun mergeStates(current: BudgetState, incoming: BudgetState): BudgetState {
        val mergedMapping = mergePredictionMapping(current.mapping, incoming.mapping)
        val mergedDescMap = mergeDescriptionMap(current.descMap, incoming.descMap)
        val mergedDescList = (current.descList + incoming.descList)
            .fold(mutableListOf<String>()) { acc, item ->
                if (acc.none { it.equals(item, ignoreCase = true) }) acc.add(item)
                acc
            }
        val mergedMonths = current.months.toMutableMap()
        incoming.months.forEach { (key, month) ->
            val sanitized = month.ensureIntegrity()
            mergedMonths[key] = sanitized
        }
        val mergedNotes = (current.notes + incoming.notes)
            .groupBy { it.id }
            .mapValues { (_, value) -> value.maxByOrNull { it.time }!! }
            .values
            .sortedByDescending { it.time }
        return current.copy(
            version = incoming.version.coerceAtLeast(current.version),
            months = mergedMonths.toSortedMap(),
            mapping = mergedMapping,
            descMap = mergedDescMap,
            descList = mergedDescList,
            notes = mergedNotes,
        )
    }

    private fun mergePredictionMapping(
        base: PredictionMapping,
        incoming: PredictionMapping
    ): PredictionMapping {
        val mergedExact = base.exact + incoming.exact
        val mergedTokens = mergeBags(base.tokens, incoming.tokens)
        return PredictionMapping(exact = mergedExact, tokens = mergedTokens)
    }

    private fun mergeDescriptionMap(
        base: DescriptionMap,
        incoming: DescriptionMap
    ): DescriptionMap {
        val mergedExact = mergeBags(base.exact, incoming.exact)
        val mergedTokens = mergeBags(base.tokens, incoming.tokens)
        return DescriptionMap(exact = mergedExact, tokens = mergedTokens)
    }

    private fun mergeBags(
        base: Map<String, Map<String, Int>>,
        incoming: Map<String, Map<String, Int>>
    ): Map<String, Map<String, Int>> {
        val result = base.toMutableMap()
        incoming.forEach { (key, bag) ->
            val merged = (result[key] ?: emptyMap()).toMutableMap()
            bag.forEach { (entry, count) ->
                merged[entry] = (merged[entry] ?: 0) + count
            }
            result[key] = merged
        }
        return result
    }

    suspend fun exportAll(): String = storage.exportAll()
}

File: ios/HomeBudgetingApp/Services/BudgetRepository.swift
import Foundation

actor BudgetRepository {
    private let storage: BudgetStorage
    private var state: BudgetState
    private var hasLoadedState = false

    init(storage: BudgetStorage = .shared) {
        self.storage = storage
        self.state = BudgetState.empty
    }

    func currentState() -> BudgetState {
        ensureLoadedState()
        return state
    }

    @discardableResult
    func refresh() -> BudgetState {
        let loaded = storage.load().ensured()
        state = loaded
        hasLoadedState = true
        return state
    }

    @discardableResult
    func createMonth(_ monthKey: String) -> BudgetState {
        updateState { state in
            guard state.months[monthKey] == nil else { return }
            let lastKey = state.months.keys.sorted().last
            var newMonth = BudgetMonth()
            if let lastKey, let lastMonth = state.months[lastKey] {
                newMonth.categories = lastMonth.categories
            }
            state.months[monthKey] = newMonth
        }
    }

    @discardableResult
    func deleteMonth(_ monthKey: String) -> BudgetState {
        updateState { state in
            state.months.removeValue(forKey: monthKey)
            state.ui.collapsed.removeValue(forKey: monthKey)
            state.ui.incomeCollapsed.removeValue(forKey: monthKey)
        }
    }

    @discardableResult
    func upsertMonth(_ monthKey: String, transform: (BudgetMonth) -> BudgetMonth) -> BudgetState {
        updateState { state in
            let month = state.months[monthKey] ?? BudgetMonth()
            state.months[monthKey] = transform(month).ensured()
        }
    }

    @discardableResult
    func setMapping(_ mapping: PredictionMapping) -> BudgetState {
        updateState { state in
            state.mapping = mapping.ensured()
        }
    }

    @discardableResult
    func setDescriptionMap(_ map: DescriptionMap) -> BudgetState {
        updateState { state in
            state.descMap = map.ensured()
        }
    }

    @discardableResult
    func setDescriptionList(_ list: [String]) -> BudgetState {
        updateState { state in
            var deduped: [String] = []
            list.forEach { value in
                let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
                guard !trimmed.isEmpty else { return }
                if !deduped.contains(where: { $0.caseInsensitiveCompare(trimmed) == .orderedSame }) {
                    deduped.append(trimmed)
                }
            }
            state.descList = deduped
        }
    }

    @discardableResult
    func setNotes(_ notes: [BudgetNote]) -> BudgetState {
        updateState { state in
            state.notes = notes.sorted { $0.time > $1.time }
        }
    }

    @discardableResult
    func setCollapsed(monthKey: String, group: String, collapsed: Bool) -> BudgetState {
        updateState { state in
            var monthMap = state.ui.collapsed[monthKey] ?? [:]
            if collapsed {
                monthMap[group] = true
            } else {
                monthMap.removeValue(forKey: group)
            }
            if monthMap.isEmpty {
                state.ui.collapsed.removeValue(forKey: monthKey)
            } else {
                state.ui.collapsed[monthKey] = monthMap
            }
        }
    }

    @discardableResult
    func setIncomeCollapsed(monthKey: String, collapsed: Bool) -> BudgetState {
        updateState { state in
            if collapsed {
                state.ui.incomeCollapsed[monthKey] = true
            } else {
                state.ui.incomeCollapsed.removeValue(forKey: monthKey)
            }
        }
    }

    @discardableResult
    func importData(_ incoming: BudgetState) -> BudgetState {
        updateState { state in
            let sanitized = incoming.ensured()
            state.version = max(state.version, sanitized.version)
            sanitized.months.forEach { key, month in
                state.months[key] = month
            }
            state.mapping = mergePredictionMapping(base: state.mapping, incoming: sanitized.mapping)
            state.descMap = mergeDescriptionMap(base: state.descMap, incoming: sanitized.descMap)
            let mergedDesc = state.descList + sanitized.descList
            var deduped: [String] = []
            mergedDesc.forEach { value in
                if !deduped.contains(where: { $0.caseInsensitiveCompare(value) == .orderedSame }) {
                    deduped.append(value)
                }
            }
            state.descList = deduped
            let mergedNotes = (state.notes + sanitized.notes)
                .reduce(into: [Int64: BudgetNote]()) { acc, note in
                    let existing = acc[note.id]
                    if let existing, existing.time >= note.time {
                        acc[note.id] = existing
                    } else {
                        acc[note.id] = note
                    }
                }
            state.notes = mergedNotes.values.sorted { $0.time > $1.time }
        }
    }

    func exportAll() -> String {
        storage.exportAll()
    }

    private func updateState(_ block: (inout BudgetState) -> Void) -> BudgetState {
        ensureLoadedState()
        var copy = state
        block(&copy)
        copy = copy.ensured()
        let sortedMonths = copy.months.sorted { $0.key < $1.key }
        copy.months = Dictionary(uniqueKeysWithValues: sortedMonths)
        state = copy
        storage.save(copy)
        return state
    }

    private func ensureLoadedState() {
        guard !hasLoadedState else { return }
        state = storage.load().ensured()
        hasLoadedState = true
    }

    private func mergePredictionMapping(base: PredictionMapping, incoming: PredictionMapping) -> PredictionMapping {
        var exact = base.exact
        incoming.exact.forEach { exact[$0.key] = $0.value }
        let tokens = mergeBags(base: base.tokens, incoming: incoming.tokens)
        return PredictionMapping(exact: exact, tokens: tokens)
    }

    private func mergeDescriptionMap(base: DescriptionMap, incoming: DescriptionMap) -> DescriptionMap {
        let exact = mergeBags(base: base.exact, incoming: incoming.exact)
        let tokens = mergeBags(base: base.tokens, incoming: incoming.tokens)
        return DescriptionMap(exact: exact, tokens: tokens)
    }

    private func mergeBags(base: [String: [String: Int]], incoming: [String: [String: Int]]) -> [String: [String: Int]] {
        var result = base
        incoming.forEach { key, bag in
            var merged = result[key] ?? [:]
            bag.forEach { entry, count in
                merged[entry] = (merged[entry] ?? 0) + count
            }
            result[key] = merged
        }
        return result
    }
}
